air/src/stack/op_flags/mod.rs

```diff
@@ -20,6 +20,15 @@ pub const NUM_DEGREE_6_OPS: usize = 16;
 /// Total number of degree 4 operations in the VM.
 pub const NUM_DEGREE_4_OPS: usize = 8;
 
+/// Total number of no shift composite flags in the VM.
+pub const NUM_NO_SHIFT_FLAGS: usize = 16;
+
+/// Total number of left shift composite flags in the VM.
+pub const NUM_LEFT_SHIFT_FLAGS: usize = 16;
+
+/// Total number of right shift composite flags in the VM.
+pub const NUM_RIGHT_SHIFT_FLAGS: usize = 16;
```

Comment:
Do these (and other) constants in this file need to be public?

Also, do we need to use 16 values for all flags? I thought we could use fewer, at least for some. We could still have all the required methods going up to 16, but the redundant ones could read values from the same places in the array.

Reply:
> Do these (and other) constants in this file need to be public? 

I'll make them private. They don't need to be public.

> Also, do we need to use 16 values for all flags? I thought we could use fewer, at least for some. We could still have all the required methods going up to 16, but the redundant ones could read values from the same places in the array.

They are still redundant. I've have populated their values with the useful ones. This way we don't need to keep track of the redundant ones with the useful ones and only need to do it in `generate_op_flags` helper function. For eg. `no_change_14` is a redundant flag and uses `no_change_12` flag values. While generating the flags, I've set the 14th index to no_change_12 flag and defined a method `no_change_14` which fetches from the 14th index of no_change_flags.

Reply:
I guess my thinking was that we won't lose any efficiency but would reduce the number of values in the array. For example, if no change 12, 13, 14, and 15 are the same, we could set `NUM_NO_SHIFT_FLAGS` to 13. And then the relative methods could work as follows:
```Rust
#[inline(always)]
pub fn no_change_12(&self) -> E {
    self.no_shift_flags[12]
}

#[inline(always)]
pub fn no_change_13(&self) -> E {
    self.no_shift_flags[12]
}

#[inline(always)]
pub fn no_change_14(&self) -> E {
    self.no_shift_flags[12]
}

#[inline(always)]
pub fn no_change_15(&self) -> E {
    self.no_shift_flags[12]
}
```